symbolic-il2cpp/src/usym.rs

```diff
@@ -106,71 +200,86 @@ impl<'a> UsymSymbols<'a> {
     /// the machine being run on.
-    pub fn parse(buf: &'a [u8]) -> Result<UsymSymbols<'a>> {
+    pub fn parse(buf: &'a [u8]) -> Result<UsymSymbols<'a>, UsymError> {
         if buf.as_ptr().align_offset(8) != 0 {
-            return Err(Error::msg("Data buffer not aligned to 8 bytes"));
+            return Err(UsymError::from(UsymErrorKind::MisalignedBuffer));
         }
         if buf.len() < mem::size_of::<raw::Header>() {
-            return Err(Error::msg("Data smaller than UsymHeader"));
+            return Err(UsymError::from(UsymErrorKind::BadHeader));
         }
-        if buf.get(..4) != Some(Self::MAGIC) {
-            return Err(Error::msg("Wrong magic number"));
+        if buf.get(..Self::MAGIC.len()) != Some(Self::MAGIC) {
+            return Err(UsymError::from(UsymErrorKind::BadMagic));
         }
 
         // SAFETY: We checked the buffer is large enough above.
         let header = unsafe { &*(buf.as_ptr() as *const raw::Header) };
         if header.version != 2 {
-            return Err(Error::msg("Unknown version"));
+            return Err(UsymError::from(UsymErrorKind::InvalidVersion));
         }
 
-        let record_count: usize = header.record_count.try_into()?;
+        let record_count: usize = header
+            .record_count
+            .try_into()
+            .map_err(|e| UsymError::new(UsymErrorKind::BadRecordCount, e))?;
+        // TODO: consider trying to just grab the records and give up on their strings if something
+        // is wrong with the string table
         let strings_offset =
             mem::size_of::<raw::Header>() + record_count * mem::size_of::<raw::SourceRecord>();
         if buf.len() < strings_offset {
-            return Err(Error::msg("Data smaller than number of records"));
+            return Err(UsymError::from(UsymErrorKind::BufferSmallerThanAdvertised));
         }
 
         // SAFETY: We checked the buffer is at least the size_of::<UsymHeader>() above.
         let first_record_ptr = unsafe { buf.as_ptr().add(mem::size_of::<raw::Header>()) };
 
-        // SAFETY: We checked the buffer has enough space for all the line records above.
+        // SAFETY: We checked the buffer has enough space for all the source records above.
         let records = unsafe {
             let first_record_ptr: *const raw::SourceRecord = first_record_ptr.cast();
             let records_ptr = ptr::slice_from_raw_parts(first_record_ptr, record_count);
             records_ptr
                 .as_ref()
-                .ok_or_else(|| Error::msg("lines_offset was null pointer!"))
+                .ok_or_else(|| UsymError::from(UsymErrorKind::BadRecords))
         }?;
 
         let strings = buf
             .get(strings_offset..)
-            .ok_or_else(|| Error::msg("No strings data found"))?;
+            .ok_or_else(|| UsymError::from(UsymErrorKind::MissingStringTable))?;
+        // TODO: null byte checking at the end of the string table?
```

Comment:
The usym string slice does not need to end in a null byte.  It's pretty difficult to check that you have to right size.

In usymlite this check exists because it uses `CStr` to get things out of the stringtable and CStr has a giant warning saying that all bets are off if you do not guarantee that whatever pointer you give it does not have a null byte before the allocation ends, you are playing with C pointers after all.  So there it's a soundness check.  Since here getting strings uses `slice::get()` there's no such soundness issues that need to be addressed here.

Reply:
thanks for the background; is there a particular reason why usym and usymlite use different strategies/types to seek into their respective string tables? is it because usyms provide the string lengths upfront?

Reply:
We only reverse-engineered these formats so I can only offer my best guesses.
I do think your question turns it upside down though, the access strategies are an effect of how the data is stored so the question is a bit answering itself.  So assuming the question why did they change the format:

- The traditional stringtable is supposed to be able to "compress" strings with the same suffix, e.g. `spamham\0` can be used for both the string "spamham" and "ham".  Since you can't do that with the format in usym maybe they did some analysis and figured this saved very little?
- NULL-terminated strings are inherently less safe as the soundness discussion above shows. All modern languages do away with them and and length-prefixed strings are basically the standard representation these days.
- I'm sort of hoping they guarantee UTF-8 encoding in the usym version of the string data.  This means they could be putting null bytes inside the strings as well.